packages/nuxt/src/app/composables/preview.ts

```diff
@@ -0,0 +1,86 @@
+import { computed, reactive, ref } from 'vue'
+import { defu } from 'defu'
+
+import { refreshNuxtData, useRoute, useRouter, useState } from '#app'
+
+interface Preview {
+  state: Record<any, unknown>
+}
+
+type GetStateFunc = (currentState: Preview['state']) => Record<any, unknown> | null | undefined | void;
+type ShouldEnableFunc = (currentState: Preview['state']) => boolean | null | undefined | void;
+
+let shouldEnablePreviewMode: ShouldEnableFunc = () => {
+  const route = useRoute()
+  const previewQueryName = 'preview'
+
+  return route.query[previewQueryName] === 'true'
+}
+
+let unregisterRefreshHook: () => any
+const _previewEnabled = ref(false)
+const previewEnabled = computed({
```

Comment:
We should handle the situation `import.meta.server`, or this will otherwise lead to a memory leak/XRSP.

Reply:
What do you mean by ?

> situation `import.meta.server`

Would `previewEnabled` be shared between requests ? Then we need to create little stub for server ?

Reply:
yes, it's very important that even if users call `usePreviewMode` on the server, that it should not be shared between requests

Reply:
Refactored composable to rely only on `useState` values, so it should not share data between requests, and it passes previous tests 🤗